snooty/types.py

```diff
@@ -82,65 +105,141 @@ def from_docutils(cls, docutils_level: int) -> "Diagnostic.Level":
+class UnexpectedIndentation(ParserDiagnostic):
+    severity = Diagnostic.Level.error
 
-    @classmethod
-    def warning(
-        cls,
-        message: str,
-        start: Union[int, Tuple[int, int]],
-        end: Union[None, int, Tuple[int, int]] = None,
-    ) -> "Diagnostic":
-        return cls.create(cls.Level.warning, message, start, end)
 
-    @classmethod
-    def error(
-        cls,
-        message: str,
-        start: Union[int, Tuple[int, int]],
-        end: Union[None, int, Tuple[int, int]] = None,
-    ) -> "Diagnostic":
-        return cls.create(cls.Level.error, message, start, end)
+class InvalidURL(ParserDiagnostic):
+    severity = Diagnostic.Level.error
+
+
+# this could captuer 'literal include expected path arg..as well?
+class ExpectedPathArg(ParserDiagnostic):
+    severity = Diagnostic.Level.error
+
+
+class ExpectedImgArg(ParserDiagnostic):
+    severity = Diagnostic.Level.error
+
+
+class ImgExpectedButNotRequired(ParserDiagnostic):
+    severity = Diagnostic.Level.warning
+
+
+class OptionsNotSupported(ParserDiagnostic):
+    severity = Diagnostic.Level.error
+
+
+class GitMergeConflictArtifactFound(ParserDiagnostic):
+    severity = Diagnostic.Level.error
+
+
+class DocUtilsParseError(ParserDiagnostic):
+    severity = Diagnostic.Level.warning
+
+
+class ErrorParsingYAMLFile(ParserDiagnostic):
+    severity = Diagnostic.Level.error
+
+
+class SemanticDiagnostic(Diagnostic):
+    pass
+
+
+class InvalidLiteralInclude(SemanticDiagnostic):
+    severity = Diagnostic.Level.error
+
+
+class SubstitutionRefError(SemanticDiagnostic):
+    severity = Diagnostic.Level.error
+
+
+# variable not declared as source constant
+class VariableNotDeclaredConstant(SemanticDiagnostic):
+    severity = Diagnostic.Level.error
+
+
+class InvalidTableStructure(SemanticDiagnostic):
+    severity = Diagnostic.Level.error
+
+
+class MissingOption(SemanticDiagnostic):
+    severity = Diagnostic.Level.error
+
+
+class GizaDiagnostic(SemanticDiagnostic):
+    pass
+
+
+class RefDiagnositc(GizaDiagnostic):
+    pass
+
+
+class MissingRef(RefDiagnositc):
+    severity = Diagnostic.Level.error
+
+
+class FailedToInheritRef(RefDiagnositc):
+    severity = Diagnostic.Level.error
+
+
+class RefAlreadyExists(RefDiagnositc):
+    severity = Diagnostic.Level.error
+
+
+class UnknownSubstitution(RefDiagnositc):
+    severity = Diagnostic.Level.warning
+
+
+class TargetDiagnostic(SemanticDiagnostic):
+    pass
+
+
+class TargetNotFound(TargetDiagnostic):
+    severity = Diagnostic.Level.error
+
+
+class AmbiguousTarget(TargetDiagnostic):
+    severity = Diagnostic.Level.error
+
+
+class DirectiveDiagnostic(SemanticDiagnostic):
+    pass
+
+
+class TodoInfo(DirectiveDiagnostic):
+    severity = Diagnostic.Level.info
+
+
+class LoadDiagnostic(Diagnostic):
+    pass
+
+
+class ErrorLoadingFile(LoadDiagnostic):
+    severity = Diagnostic.Level.error
+
+
+class OSDiagnostic(LoadDiagnostic):
+    pass
+
+
+class CannotOpenFile(OSDiagnostic):
```

Comment:
Let's actually mock out making this a little more interesting!

```
class CannotOpenFile(OSDiagnostic):
    def __init__(self, path: Path, reason: str, start: ..., end: ...) -> None:
        super().__init__(self, "Error opening {str(path)}: {reason}", start, end)
        self.path = path
        self.reason = reason
```

Then, you don't need to construct the reason each time it's instantiated: you just construct e.g. `CannotOpenFile(parent_identifier.file, os.strerror(errno.ENOENT), start)` or `CannotOpenFile(self.asset.fileid, err.strerror, start)`.

How does this sound to you?